feat(frontend,cms): integrate real API data for My Reading page

ericsen-tsai opened the pull request:
## Summary

This PR integrates the My Reading page with the backend API, replacing mock data with real GraphQL queries. It adds a new GraphQL query `getMemberPostsWithAnswers` in the CMS backend and implements the frontend API integration to fetch and display member posts with their associated essay and choice answers.

## Changes

- **CMS Backend:**
  - Added new GraphQL query `getMemberPostsWithAnswers` in `extend-schema.ts` that fetches posts with pagination and includes all related essay and choice answers for a given member
  - Added TypeScript types (`EssayAnswerQueryResult`, `ChoiceAnswerQueryResult`, `PostQueryResult`, `MemberPostWithAnswers`) in `types/extend-schema.ts` for type safety
  - Updated GraphQL schema to include the new query and response types
  - Query supports pagination with `take` and `skip` parameters and returns posts sorted by last answered time

- **Frontend:**
  - Created API layer for the new GraphQL query (`api/extended.ts`, `api/graphql/extended.ts`)
  - Added React Query hook `useGetMemberPostsWithAnswersQuery` for data fetching with caching
  - Updated `my-reading/index.tsx` to use real API data instead of mock data
  - Removed mock data file (`mock-data.ts`)
  - Added utility function `parseMemberPostsWithAnswersToPostQuestionAnswers` to transform API response to component format
  - Implemented loading states in the UI
  - Added pagination prefetching for better UX (prefetches next page when user is on current page)

## Additional Notes

- The query aggregates answers from both `PostEssayAnswer` and `PostChoiceAnswer` to determine which posts a member has answered
- Posts are sorted by `lastAnsweredTime` (most recently answered first)
- The implementation includes proper error handling and logging for debugging
- Pagination is handled client-side with React Query for optimal performance

## Screenshot(s)

## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/CMS-2a28dbdca932802c8974eab93d6c5e68?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)

ericsen-tsai committed:
feat(graphql): add getMemberPostsWithAnswers query to fetch member posts with associated answers

- Introduced a new GraphQL query `getMemberPostsWithAnswers` that retrieves posts associated with a specific member, including pagination support.
- The query aggregates unique post IDs from both essay and choice answers, returning the posts along with their respective answers.

ericsen-tsai committed:
feat(frontend): enhance My Reading component with data fetching and loading state

- Integrated `useGetMemberPostsWithAnswersQuery` to fetch member posts with answers, supporting pagination.
- Added loading state handling in the `PostQuestionAnswersList` component.
- Removed mock data dependency and streamlined the component to utilize real API data.

fix(frontend,cms): optimize getMemberPostsWithAnswers query and answer order

feat(frontend): add access token to getMemberPostsWithAnswers query

ericsen-tsai committed:
feat(frontend): adjust prefetch next page logic and query key

ericsen-tsai committed:
fix(cms): remove unnecessary fields from getMemberPostsWithAnswers query

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -336,6 +342,270 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                  new Date(a.lastAnsweredTime).getTime()
+                )
+              })
+              .slice(skip ?? 0, (skip ?? 0) + (take ?? 5))
```
thought:
一般來說，我會盡量避免寫複雜的 raw sql statement，因為不易閱讀和維護，所以盡量選擇 js 版本的寫法。
然而，目前 js 版本的解法，看起來也是挺複雜的，因為需要 query 多個 tables，自己排序和做 pagination。

除了易讀性之外，raw sql statement 的效能比 js 實作來得好。

考量上述兩點，看要不要考慮寫 raw sql statement 來拿取資料呢？

例如：
```
   WITH per_type AS (
      /* For each post, find the latest choice answer from this member */
      SELECT q."post" AS post_id, MAX(a."createdAt") AS answered_at
      FROM "PostChoiceAnswer" a
      JOIN "PostChoiceQuestion" q ON q.id = a."question"
      WHERE a."member" = :member_id AND q."post" IS NOT NULL
      GROUP BY q."post"

      UNION ALL

      /* For each post, find the latest essay answer from this member */
      SELECT q."post" AS post_id, MAX(a."createdAt") AS answered_at
      FROM "PostEssayAnswer" a
      JOIN "PostEssayQuestion" q ON q.id = a."question"
      WHERE a."member" = :member_id AND q."post" IS NOT NULL
      GROUP BY q."post"
    ),
    /* For each post, find the latest answer from this member */
    latest AS (
      SELECT post_id, MAX(answered_at) AS last_answered_at
      FROM per_type
      GROUP BY post_id
    )
    /* 根據回答時間的優先排序文章，並且做 cursor-based pagination */
    SELECT p.id, p.slug, p.title, latest.last_answered_at AS answered_at_cursor
    FROM latest
    JOIN "Post" p ON p.id = latest.post_id
    WHERE 
       /* 第一次 query，所以沒有 cursor */
       :cursor_answered_at IS NULL 
       /* 根據回答時間排序 */
       OR latest.last_answered_at < :cursor_answered_at
       /* 如果回答時間相同，根據文章的 id 做 tie-breaker */
       OR (latest.last_answered_at = :cursor_answered_at AND latest.post_id < :cursor_post_id)
    ORDER BY latest.last_answered_at DESC, latest.post_id DESC
    LIMIT :limit;
  ```

ericsen-tsai replied:
You're right, since the performance of graphql query is way slower than raw sql, raw sql is the better solution for this case.
Modified in [refactor(cms,frontend): update getMemberPostsWithAnswers query to use cursor-based pagination](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/91acd84e09e065aae21409520ada0cf52c40ac5a)

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/post-question-answers-list/index.tsx:
```diff
@@ -0,0 +1,104 @@
+    >
+      {postQuestionAnswers.map((group) => {
+        return (
+          <AccordionItem
+            key={group.title}
+            value={group.title}
```
Replied in https://github.com/kids-reporter/kids-reporter-monorepo/pull/825#discussion_r2513397905

ericsen-tsai commented on packages/frontend/src/globals.css:
```diff
@@ -465,9 +455,9 @@
 
-/* Swei Marker Sans Font Variants */
 @utility font-swei {
-  font-family: var(--font-family-swei);
+  font-family: var(--font-swei-marker);
+  font-weight: 700;
 }
```
Fixed in other PR comment [fix(frontend): fix copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/825/commits/94f3accb0bb40d93e94265c8a4e06df96949766a)

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/post-question-answers-list/choice-answer-item.tsx:
```diff
@@ -0,0 +1,55 @@
+          <div className="text-white">
+            {answer.correct ? <CorrectIcon /> : <IncorrectIcon />}
+          </div>
+        </div>
+        <div className="flex flex-1 items-center rounded-r-xl bg-white px-4 py-4">
+          <p className="prose-p1-medium text-neutral-900">
+            {`${(answer.choiceIndex ?? 0) + 1}. ${selectedOption?.content}`}
+          </p>
```
Fixed in other PR comment [fix(frontend): fix copilot comments](https://github.com/kids-reporter/kids-reporter-monorepo/pull/825/commits/94f3accb0bb40d93e94265c8a4e06df96949766a)

ericsen-tsai committed:
fix(frontend): update font family and improve query invalidation logic

- Changed font family from `font-family-swei` to `font-swei` in related articles and My Reading components.
- Enhanced `useBatchSubmitAnswers` hook to invalidate additional queries for better data consistency.
- Refactored date formatting logic in `PostQuestionAnswersList` to simplify condition checks.

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/index.tsx:
```diff
@@ -0,0 +1,121 @@
+  useEffect(() => {
+    if (typeof totalCount === 'number' && memberPostsWithAnswersData) return
+    setTotalCount(memberPostsWithAnswersData?.totalCount)
+  }, [memberPostsWithAnswersData, totalCount])
```
Since user only can add answer on the other page, the total page should be static.
I added invalidation for member post answers instead.

ericsen-tsai replied:
Fixed in [fix(frontend): update font family and improve query invalidation logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/65c2167bf6c78efc45f7db3c5e8742e2cd634915)

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/post-question-answers-list/index.tsx:
```diff
@@ -0,0 +1,104 @@
+  isLoading: boolean
+}
+
+function formatDate(dateString: string): string {
+  const date = new Date(dateString)
+  if (!date || isNaN(date.getTime())) {
+    return ''
+  }
```
Fixed in [fix(frontend): update font family and improve query invalidation logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/65c2167bf6c78efc45f7db3c5e8742e2cd634915)

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/index.tsx:
```diff
@@ -0,0 +1,121 @@
+          <MembershipSideMenu />
+        </div>
+        <div className="flex flex-1 flex-col px-6 tablet:px-8 desktop:px-0">
+          <h1 className="mb-6 prose-h4-large font-family-swei text-neutral-900">
```
Fixed in [fix(frontend): update font family and improve query invalidation logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/65c2167bf6c78efc45f7db3c5e8742e2cd634915)

ericsen-tsai commented on packages/frontend/src/modules/membership/my-reading/post-question-answers-list/essay-answer-item.tsx:
```diff
@@ -0,0 +1,21 @@
+import { ThumbUpIcon } from '@/icons/miscellaneous'
+
+function EssayAnswerItem({ answer }: { answer: PostEssayAnswer }) {
+  const likesCount = (answer.likesCount ?? 0) > 99 ? '99+' : answer.likesCount
```
Fixed in [fix(frontend): update font family and improve query invalidation logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/65c2167bf6c78efc45f7db3c5e8742e2cd634915)

ericsen-tsai committed:
refactor(cms,frontend): update getMemberPostsWithAnswers query to use cursor-based pagination

- Modified the `getMemberPostsWithAnswers` query to use raw sql with a `cursor` for improved pagination.
- Updated related frontend components to utilize the new cursor-based pagination, including changes to the API hooks and My Reading component.
- Adjusted types and GraphQL queries to reflect the new pagination logic, ensuring compatibility with existing data structures.

ericsen-tsai commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -336,6 +342,270 @@ export const extendGraphqlSchema = graphql.extend(() => {
+                        allAnswers[0].updatedAt ?? allAnswers[0].createdAt ?? ''
+                      ).toISOString()
+                    : new Date().toISOString()
+                )
```
I changed this logic to raw sql.

[refactor(cms,frontend): update getMemberPostsWithAnswers query to use cursor-based pagination](https://github.com/kids-reporter/kids-reporter-monorepo/pull/826/commits/91acd84e09e065aae21409520ada0cf52c40ac5a)

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -341,77 +342,119 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              FROM "Post" p
+              INNER JOIN post_last_answered pla ON pla.post_id = p.id
+              ${whereClause}
+              ORDER BY pla.last_answered_time DESC
```
question:
使用 cursor 的話，如果有兩篇 post 的 last_answered_time 是一樣的話，需要 tie-breaker 嗎？

ericsen-tsai replied:
I think in real scenario, it's really hard to have one user that answer multiple posts at the same time.
So I think maybe this checking logic can be skipped for now.

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -341,77 +342,119 @@ export const extendGraphqlSchema = graphql.extend(() => {
-              return { posts: [], totalCount: 0 }
+          try {
+            const whereClause = cursor
+              ? Prisma.sql`WHERE pla.last_answered_time < ${cursor}::timestamp`
```
question:
是不是要用 `WHERE pla.last_answered_time <= ${cursor}::timestamp` ?
如果有相同的 last_answered_time 的話，這樣子才不會漏拿 post。

不過，如果改成 `pla.last_answered_time <= ${cursor}::timestamp` 的話，
下面 sql statement 可能要加上 `OFFSET 1`，去 skip 掉 cursor 的那筆 record。

ericsen-tsai replied:
Replied in https://github.com/kids-reporter/kids-reporter-monorepo/pull/826#discussion_r2517525895. Thanks!

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -341,77 +342,119 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              INNER JOIN post_last_answered pla ON pla.post_id = p.id
+              ${whereClause}
+              ORDER BY pla.last_answered_time DESC
+              LIMIT ${takeValue + 1}
```
question:
想問問 LIMIT 故意多拿一筆的原因是？

ericsen-tsai replied:
If I want to check if there's next page, I can just use this query with `const hasNextPage = postsWithLastAnsweredTime.length > takeValue` instead of another query to count the post.

nickhsine replied:
This is brilliant!

nickhsine commented on packages/cms/graphql/extend-schema.ts:
```diff
@@ -458,51 +501,35 @@ export const extendGraphqlSchema = graphql.extend(() => {
+              )
+              const choiceAnswers = choiceAnswersData.filter(
+                (a) => a.question?.post?.id?.toString() === postId
+              )
```
nitpick:
從 gql 拿回來的 `essayAnswersData` 和 `choiceAnsweresData` 是不是不用再 filter 了。

ericsen-tsai replied:
It's still necessary since `essayAnswersData` and `choiceAnsweresData` are for all posts returned from sql

nickhsine commented on packages/frontend/src/api-utils/react-query/hooks/extended.ts:
想問問，file name 用 `extended` 是有什麼特殊含義嗎？
是指裡面的 gql mutation 或是 query 是 call custom GQL 的意思嗎？

ericsen-tsai replied:
Yes, I tried to separate hooks and graphql to align list name / extend-schema in `cms` to track the api call pipeline

nickhsine reviewed (commented):
LGTM 👍 💯 🥇
